File: october18_eventual_safe_states.py
#User function Template for python3

from typing import List
from collections import deque
class Solution:    
    def eventualSafeNodes(self, V : int, adj : List[List[int]]) -> List[int]:
        # code here
        ans=[]
        arr=[[] for _ in range(V)]
        indegree=[0]*V
        for i in range(V):
            for it in adj[i]:
                arr[it].append(i)
                if indegree[i]==0:
                    indegree[i]=1
                else:
                    indegree[i]+=1
        q=deque()
        for i in range(V):
            if indegree[i]==0:
                q.append(i)
        
        while q:
            node=q.popleft()
            ans.append(node)
            for it in arr[node]:
                indegree[it]-=1
                if indegree[it]==0:
                    q.append(it)
        ans.sort()
        return ans

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__=="__main__":
    T = int(input())
    for t in range(T):
        
        V, E = map(int, input().strip().split())
        adj = [[] for i in range(V)]
        for i in range(E):
            u, v = map(int, input().strip().split())
            adj[u].append(v)
        obj = Solution()
        ans = obj.eventualSafeNodes(V, adj)
        for nodes in ans:
            print(nodes, end = ' ')
        print()
            


# } Driver Code Ends

File: october18_eventual_safe_states.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++


class Solution
{
public:
    vector<int> eventualSafeNodes(int V, vector<int> adj[])
    {
        // code here
        vector<int> ans;
        vector<int> arr[V];
        int indegre[V] = {0};
        for (int i = 0; i < V; i++)
        {
            for (auto it : adj[i])
            {
                arr[it].push_back(i);
                indegre[i]++;
            }
        }
        queue<int> q;
        for (int i = 0; i < V; i++)
        {
            if (indegre[i] == 0)
                q.push(i);
        }
        while (!q.empty())
        {
            int node = q.front();
            q.pop();
            ans.push_back(node);
            for (auto it : arr[node])
            {
                indegre[it]--;
                if (indegre[it] == 0)
                    q.push(it);
            }
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin >> t;
    while (t--)
    {

        int V, E;
        cin >> V >> E;
        vector<int> adj[V];

        for (int i = 0; i < E; i++)
        {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
        }

        Solution obj;
        vector<int> safeNodes = obj.eventualSafeNodes(V, adj);
        for (auto i : safeNodes)
        {
            cout << i << " ";
        }
        cout << endl;
    }
}

// } Driver Code Ends